### PR TITLE
Frontend: Update to Eslint 9

### PR DESCRIPTION
**Pull request by jackw**
**What is this feature?**

This PR updates the codebase to use Eslint 9.

Massive thanks to @ashharrison90 for the support/answering all my random questions. 🏅 

**Why do we need this feature?**

So we stay up to date with dependencies and stop using a deprecated version of eslint.

**Who is this feature for?**

Grafana devs / contributors

**Which issue(s) does this PR fix?**:

Related: grafana/plugin-tools/issues/879

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/contribute/release-notes/#how-to-determine-if-content-belongs-in-whats-new), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/contribute/release-notes/) doc.

### CODE REVIEW

File: package.json
Comment (tomratcliffe) [inline package.json:431]: (If you're testing this locally, don't forget to delete this before running `yarn link ...`)

File: eslint.config.js
Comment (tomratcliffe) [inline eslint.config.js:263]: The rules aren't all being applied here - `testing-library` needs compatibility changes in order to work, and the rules need to be spread differently. I've made the changes locally and can add to this PR if you're happy for me to do so.

Might be worth us double checking that rules will still fail in the different places that used to have separate configs

**Review by tomratcliffe — COMMENTED** [review]
I love it 🎉 

This also makes it a lot easier for us to allow devs to have the betterer rules show up in their IDEs, by hooking into the config ini file and optionally including additional rules 🙌  I can do that in a follow up PR once this one is merged

File: .eslintignore
Comment (jackw) [reply .eslintignore:1]: You can go ahead and delete `public/lib/monaco`, it was removed from the codebase quite a while ago. Even so it might be best to add this back if we feel it's gonna cause bother for folk. 👍

File: eslint.config.js
Comment (jackw) [reply eslint.config.js:263]: Please feel free to push to the PR @tomratcliffe. I'd appreciate the assistance as it's already been more time consuming than I thought to get this far. 😅

File: package.json
Comment (jackw) [reply package.json:431]: It was more to clean the lockfile until the tsconfig PR is merged. I'm hoping CI won't let us merge this like this. 🤞😄

File: eslint.config.js
Comment (tomratcliffe) [inline eslint.config.js:3]: Actually, 40 minutes ago at the time of writing, a maintainer said that a PR that will help add compatibility for the eslint plugin will be reviewed/merged soon! 
So might be worth us checking the status of [this PR](https://github.com/testing-library/eslint-plugin-testing-library/pull/925#issuecomment-2418896886) before we merge 🙌

Comment (jackw) [reply eslint.config.js:3]: @tomratcliffe the PR was [merged 10hrs ago](https://github.com/testing-library/eslint-plugin-testing-library/pull/925#issuecomment-2420735710). I'm happy for us to get this out the door first and then bump it and clean up.

**Review by academo — COMMENTED** [review]
ran locally and tested it with neovim and eslint LSP server. Works well detecting issues and autofixing just as the current implementation

**Review by academo — APPROVED** [review]

File: package.json
Comment (tomratcliffe) [inline package.json:39]: I pushed this as well to make sure that the new `eslint.config.js` file is caught by prettier

**Review by tomratcliffe — APPROVED** [review]
💯 Great work, thank you! 

I've checked that the alerting custom rules still apply, and all seems to be working :+1: I did see one case where the `eslintcache` was stuck and was reporting on an issue that I'd fixed, but I haven't seen it again since so I think it was just related to me having jumped branches/reinstalled

I'll add the `structured-clone` fixes in a separate branch as there are ones to move from alerting as well

🚀

File: packages/grafana-eslint-rules/tests/no-aria-label-e2e-selectors.test.js
Comment (tomratcliffe) [reply packages/grafana-eslint-rules/tests/no-aria-label-e2e-selectors.test.js:5]: https://github.com/grafana/grafana/pull/95406

Comment (tomratcliffe) [reply packages/grafana-eslint-rules/tests/no-aria-label-e2e-selectors.test.js:5]: I've merged that PR now, so you should be able to remove the manual imports

**Review by ashharrison90 — COMMENTED** [review]
what's with the modifications to `.go` files? 🤔

**Comment by itsmylife** [comment]
> what's with the modifications to `.go` files? 🤔

You don't need to specify the type again while you are defining the items in map. The type can be inferred from the variable definition. They are unnecessary. And somehow maybe eslint, or some other tool triggered a cleanup. So we have these changes.

**Comment by ashharrison90** [comment]
> > what's with the modifications to `.go` files? 🤔
> 
> You don't need to specify the type again while you are defining the items in map. The type can be inferred from the variable definition. They are unnecessary. And somehow maybe eslint, or some other tool triggered a cleanup. So we have these changes.

so that might be great, might not be. but that's what i'm worried about 😬 if `eslint` is now looking at `.go` files that's a change in behaviour we should probably consider separately to this change

**Comment by itsmylife** [comment]
@ashharrison90 I agree that we should remove those `.go` changes from this PR. And yeah I'd like to know too :)

**Comment by jackw** [comment]
> what's with the modifications to `.go` files? 🤔

A fine question! It appears to have happened when I merged main into this branch. 😕

**Comment by jackw** [comment]
> what's with the modifications to `.go` files? 🤔

It's caused by the git hook `backend-format` and the merge of main. My IDE complains about it:

image

Switching to main and running `gofmt -w -s pkg/services/authz/zanzana/common/info.go` will also trigger it. 

As there is [no go format check in CI](https://raintank-corp.slack.com/archives/CMSUYD6H3/p1711553314147999?thread_ts=1711530644.579229&cid=CMSUYD6H3) it will only surface when someone touches the file and uses the git hooks.

I'll revert them. 👍

**Review by ashharrison90 — APPROVED** [review]
lgtm 👍

File: packages/grafana-eslint-rules/tests/no-aria-label-e2e-selectors.test.js
Comment (jackw) [reply packages/grafana-eslint-rules/tests/no-aria-label-e2e-selectors.test.js]: Sorted! Thanks for handling that in a separate PR @tomratcliffe 👏

File: package.json
Comment (ashharrison90) [inline package.json:94]: do we actually use this anywhere? 🤔

Comment (jackw) [reply package.json:94]: It's a peer dependency of the [grafana/eslint-config](https://github.com/grafana/eslint-config-grafana/blob/master/base.js#L63-L72).

**Pull request merged**
